src/app/components/login/login.component.ts

```diff
@@ -15,50 +16,65 @@ export class MyErrorStateMatcher implements ErrorStateMatcher {
   ): boolean {
     const isSubmitted = form && form.submitted;
     return !!(
-      control &&
-      control.invalid &&
+      control?.invalid &&
       (control.dirty || control.touched || isSubmitted)
     );
   }
 }
 
-const pattern = '^[a-zA-Z0-9]*$';
+const validCharacters = '^[a-zA-Z0-9]*$';
```

Comment:
```suggestion
const alphaNumeric = '^[a-zA-Z0-9]*$';
```

Look for the custom Angular Validators if u have this already, to save u from declaring the pattern